src/blog/articles/javascript-async-iterators/index.md

```diff
@@ -0,0 +1,689 @@
+}
+```
+
+In the example above, we are iterating over an array using the `for...of` syntax. If we run this code, this is what we will get as output:
+
+```text
+Driulis Gonzalez Morales
+Ilias Iliadis
+Tadahiro Nomura
+Anton Geesink
+Teddy Riner
+Ryoko Tani
+```
+
+The same syntax works also for iterating over the characters of a string:
+
+```javascript
+const judoka = 'Ryoko Tani'
+
+for (const char of judoka) {
+  console.log(char)
+}
+```
+
+This will print:
+
+```text
+R
+y
+o
+k
+o
+
+T
+a
+n
+i
+```
+
+And we can even use this for `Set` and `Map`:
+
+```javascript
+const medals = new Set(['gold', 'silver', 'bronze'])
+
+for (const medal of medals) {
+  console.log(medal)
+}
+```
+
+Which is going to output:
+
+```text
+gold
+silver
+bronze
+```
+
+`Map` is especially interesting because we can use _destructuring_ to iterate over key-value pairs:
+
+```javascript
+const medallists = new Map([
+  ['Teddy Riner', 33],
+  ['Driulis Gonzalez Morales', 16],
+  ['Ryoko Tani', 16],
+  ['Ilias Iliadis', 15]
+])
+
+for (const [judoka, medals] of medallists) {
+  console.log(`${judoka} has won ${medals} medals`)
+}
+```
+
+This will output:
+
+```text
+Teddy Riner has won 33 medals
+Driulis Gonzalez Morales has won 16 medals
+Ryoko Tani has won 16 medals
+Ilias Iliadis has won 15 medals
+```
+
+Finally, if you want to iterate over the key-value pairs of a regular object using the `for...of` syntax, we can do that by using the helper `Object.entries`:
+
+```javascript
+const medallists = {
+  'Teddy Riner': 33,
+  'Driulis Gonzalez Morales': 16,
+  'Ryoko Tani': 16,
+  'Ilias Iliadis': 15
+}
+
+for (const [judoka, medals] of Object.entries(medallists)) {
+  console.log(`${judoka} has won ${medals} medals`)
+}
+```
+
+This will produce the same output as the previous example.
+
+What's interesting here is that, if we try to use the `for...of` syntax directly on the object `medallists` (without `Object.entries`), we get the following error:
+
+```text
+for (const [judoka, medals] of medallists) {
+                               ^
+
+TypeError: medallists is not iterable
+    at Object.<anonymous> (.../05-for-of-object.js:8:32)
+    at Module._compile (node:internal/modules/cjs/loader:1108:14)
+    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)
+    at Module.load (node:internal/modules/cjs/loader:988:32)
+    at Function.Module._load (node:internal/modules/cjs/loader:828:14)
+    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
+    at node:internal/main/run_main_module:17:47
+```
+
+Let's read this error once again: `medallists is not iterable`!
+
+Yeah, the error is clear: a regular JavaScript object is not _iterable_, while arrays, strings, maps, and sets are!
+
+But what does it mean for an object to be _iterable_?
+
+During the rest of this article, we will learn how JavaScript knows if a given object is **iterable** and how we can make our own custom _iterable_ objects.
```

Comment:
I'd add at the end:

"But first let's quickly take a look at how we can use *async* iterators"